engine/src/main/java/org/pentaho/di/job/entries/shell/JobEntryShell.java

```diff
@@ -476,21 +477,28 @@ private void executeShell( Result result, List<RowMetaAndData> cmdRows, String[]
           fileObject = createTemporaryShellFile( tempFile, realScript );
         }
       } else {
-        if ( insertScript ) {
-          realScript = environmentSubstitute( script );
-          tempFile = KettleVFS.createTempFile( KETTLE, "shell", System.getProperty( TEMP_DIR ), this );
+        if ("Y".equalsIgnoreCase( System.getProperty( Const.KETTLE_EXECUTE_TEMPORARY_GENERATED_FILE, "Y" ) )) {
+          if ( insertScript ) {
+            realScript = environmentSubstitute( script );
+            tempFile = KettleVFS.createTempFile( KETTLE, SHELL, System.getProperty( TEMP_DIR ), this );
+          } else {
+            String realFilename = environmentSubstitute( getFilename() );
+            URI uri = new URI( realFilename );
+            realFilename = uri.getPath();
+            try ( FileInputStream fis = new FileInputStream( realFilename ) ) {
+              realScript = IOUtils.toString( fis, "UTF-8" );
+            }
+            // PDI-19676 - creating a temp file in same file location to avoid script failure.
+            String parentDir = Paths.get( realFilename ).getParent().toString();
+            tempFile = KettleVFS.createTempFile( KETTLE, SHELL, parentDir, this );
+          }
+          fileObject = createTemporaryShellFile( tempFile, realScript );
         } else {
-          String realFilename = environmentSubstitute( getFilename() );
-          URI uri = new URI( realFilename );
-          realFilename = uri.getPath();
-          try ( FileInputStream fis = new FileInputStream( realFilename ) ) {
-            realScript = IOUtils.toString( fis, "UTF-8" );
+          if ( insertScript ) {
+            tempFile = KettleVFS.createTempFile( "kettle", SHELL, System.getProperty( "java.io.tmpdir" ), this );
```

Comment:
Please make use of the already existing 'KETTLE' and 'TEMP_DIR' variables instead of the string values.